src/state/dashboard/reducer.js

```diff
@@ -0,0 +1,179 @@
+    case 'DASHBOARD_TIME_REQUESTS_SUCCESS': {
+      const { samplesOverTime, experimentsOverTime, jobs } = action.data;
+      return {
+        ...state,
+        samplesOverTime,
+        experimentsOverTime,
+        jobs,
+        isLoading: false
+      };
+    }
+    case 'DASHBOARD_TIME_OPTIONS_UPDATED': {
+      const { timeOptions } = action.data;
+      return {
+        ...state,
+        timeOptions,
+        isLoading: true
+      };
+    }
+    default: {
+      return state;
+    }
+  }
+};
+
+export default dashboardReducer;
+
+// chart selectors for creating chart data for individual charts on dashboard
+
+export function getTotalLengthofQueuesByType(state) {
+  const stats = state.dashboard.stats;
+  return Object.keys(stats).map(jobType => {
+    return {
+      name: jobType.split('_')[0],
+      value: stats[jobType].open + stats[jobType].pending
+    };
+  });
+}
+
+export function getJobsByStatus(state) {
+  const stats = state.dashboard.stats;
+  return Object.keys(stats).reduce((accum, jobType) => {
+    accum[jobType] = Object.keys(stats[jobType]).reduce(
+      (allStatuses, status) => {
+        if (status !== 'total' && status !== 'average_time') {
+          allStatuses.push({
+            name: status,
+            value: stats[jobType][status]
+          });
+        }
+        return allStatuses;
+      },
+      []
+    );
+    return accum;
+  }, {});
+}
+
+function convertSecToMinHours(sec) {
+  const hours = Math.floor(sec / 3600),
+    minutes = Math.floor((sec % 3600) / 60);
+  if (isNaN(hours) || isNaN(minutes)) {
+    return `N/A`;
+  } else {
+    return `${hours} hr ${minutes} min`;
+  }
+}
+
+export function getAllEstimatedTimeTilCompletion(state, jobType) {
+  const stats = state.dashboard.stats;
+
+  if (!Object.keys(stats).length) return {};
+
+  return Object.keys(stats).reduce((allEstimatedTimes, jobType) => {
+    const estimateSec =
+      (stats[jobType].open + stats[jobType].pending) *
+      parseInt(stats[jobType].average_time, 10);
+    // we're assuming that average_time is in seconds...
+    allEstimatedTimes[jobType] = convertSecToMinHours(estimateSec);
+    return allEstimatedTimes;
+  }, {});
+}
+
+export function getExperimentsCount(state) {
+  const { experiments: { count = 0 } } = state.dashboard;
+
+  return count;
+}
+
+export function getSamplesCount(state) {
+  const { samples: { count = 0 } } = state.dashboard;
+
+  return count;
+}
+
+export function getJobsCompletedOverTime(state) {
+  const { jobs, timeOptions: { timePoints } } = state.dashboard;
+
+  return timePoints.reduce((jobTotals, time, i) => {
+    const dataPoint = {
+      date: time.utc().format()
+    };
+
+    Object.keys(jobs).forEach(jobName => {
+      if (jobs[jobName].all.length === timePoints.length) {
+        // accumulate count completed over time
+        const jobCount =
+          i === 0
+            ? jobs[jobName].all[i].count
+            : jobTotals[i - 1][jobName] + jobs[jobName].all[i].count;
```

Comment:
Can you explain to me what this addition means? What about the `jobName` field of the last element in `jobTotals`?

Reply:
So, if this is the first date in the time range, `jobCount` will just be equal to the `count` for that date. If not, I accumulate the total of jobs run from the first date in the time range up to this current date for `jobCount`.